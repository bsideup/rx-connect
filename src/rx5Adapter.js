import { Observable, Subject, BehaviorSubject, ReplaySubject } from "rxjs-v5";
import "rxjs-v5/add/observable/of";
import "rxjs-v5/add/observable/interval";
import "rxjs-v5/add/observable/merge";
import { scan, debounce, map, pluck, take } from "rxjs-v5/operators";
import ObservableSymbol from "symbol-observable";

const handler = {
  get: function(obj, key) {
    return obj.hasOwnProperty(key) ? obj[key] : Observable;
  }
};

const ObservableProxy = new Proxy({
  scan,
  debounce,
  map,
  pluck,
  take,
  interval: Observable.interval,
  merge: Observable.merge,
  of: Observable.of
}, handler);

export default {

    Rx: {
      Observable: ObservableProxy,
      Subject,
      ReplaySubject,
      BehaviorSubject,
    },

    next(o, value) {
        o.next(value);
    },

    isObservable(obj) {
        return obj && typeof obj[ObservableSymbol] === 'function';
    },

    unsubscribe(subscription) {
        subscription.unsubscribe();
    }
}
