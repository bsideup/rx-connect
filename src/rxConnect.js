import React from "react";
import isPlainObject from "lodash.isplainobject";
import isObject from "lodash.isobject";
import isArray from "lodash.isarray";

const DEFAULT_OPTIONS = {
    noDebounce: false
};

function getDisplayName(WrappedComponent) {
    return WrappedComponent.displayName || WrappedComponent.name || "Component";
}
function isObservable(obj) {
    if (!obj) {
        return false;
    }

    return getAdapter().isObservable(obj);
}

export function getAdapter() {
    const adapter = rxConnect.adapter || require("./rx6adapter");
    return adapter.__esModule ? adapter.default : adapter;
}

export default function rxConnect(selector, options = DEFAULT_OPTIONS) {
    return WrappedComponent => class RxConnector extends React.PureComponent {

        static displayName = "RxConnector";

        stateSubscription = undefined;

        shouldDebounce = false;

        subProps = {};

        constructor(props) {
            super(props);
            const { Rx } = getAdapter();

            this.props$ = new (Rx.BehaviorSubject)(props);
        }

        componentWillMount() {
            const { Rx: { Observable } } = getAdapter();

            this.shouldDebounce = false;

            let mutations$ = selector;
            if(!isObservable(mutations$)) {
                if (typeof selector === 'function') {
                    mutations$ = selector(this.props$);
                } else {
                    // eslint-disable-next-line no-console
                    console.error(`Selector must be a function or an Observable. Check rxConnect of ${getDisplayName(WrappedComponent)}. Got: `, selector);
                    return;
                }
            }

            if(!isObservable(mutations$)) {
                    // eslint-disable-next-line no-undef
                if (mutations$ && typeof mutations$[Symbol.iterator] === 'function') {
                    mutations$ = Observable.merge(...mutations$);
                } else {
                    // eslint-disable-next-line no-console
                    console.error(`Selector must return an Observable or an iterator of observables. Check rxConnect of ${getDisplayName(WrappedComponent)}. Got: `, mutations$);
                    return;
                }
            }
            
            this.stateSubscription = mutations$.pipe(
                Observable.scan((state, mutation) => {
                    if (typeof mutation === "function") {
                        return mutation(state);
                    }

                    if (isPlainObject(mutation)) {
                        return Object.assign({}, state, mutation);
                    }

                    if (isObject(mutation) && !isArray(mutation)) {
                        return Object.assign({}, state, {...mutation});
                    }

                    // eslint-disable-next-line no-console
                    console.error(`Mutation must be a plain object or function. Check rxConnect of ${getDisplayName(WrappedComponent)}. Got: `, mutation);
                    return state;
                }, {}),
                Observable.debounce(() => (!options.noDebounce && this.shouldDebounce) ? Observable.interval(0) : Observable.of()),
                ).subscribe(state => {
                  this.subProps = state;
                  this.forceUpdate();
                });
        }

        componentDidMount() {
            this.shouldDebounce = true;
        }

        componentWillUnmount() {
            getAdapter().unsubscribe(this.stateSubscription);
        }

        componentWillReceiveProps(nextProps) {
            getAdapter().next(this.props$, nextProps);
        }

        render() {
            return React.createElement(WrappedComponent, this.subProps, this.props.children);
        }
    };
}
