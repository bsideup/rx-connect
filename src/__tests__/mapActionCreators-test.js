import { mapActionCreators, rxConnect } from "../";
import rx5Adapter from "../rx5Adapter";
import { getAdapter } from "../rxConnect";

const suites = {
    "RxJS 6": () => {},
    "RxJS 5": () => rxConnect.adapter = rx5Adapter,
}

Object.entries(suites).forEach(([ name, initializer ]) => describe(name, () => {
    const { Rx } = getAdapter();
    beforeEach(() => {
        initializer();
    });

    it("passes non-observables as is", async () => {
        const props = await mapActionCreators({ a: 123, b: "hi!" }).toPromise();

        expect(props).toMatchSnapshot();
    });

    it("strips dollar sign from Observable property names", async () => {
        const actions = {
            a$: new Rx.Subject()
        };

        const props = await mapActionCreators(actions).toPromise();

        expect(props).toMatchSnapshot();
    });

    it("creates FuncSubject-like action", async () => {
        const actions = {
            a$: new Rx.BehaviorSubject()
        };

        const props = await mapActionCreators(actions).toPromise();

        props.a(1, 2, 3);

        expect(actions.a$.getValue()).toMatchSnapshot();
    });

}));
