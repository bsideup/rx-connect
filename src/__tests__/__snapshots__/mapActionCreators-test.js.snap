exports[`RxJS 5 creates FuncSubject-like action 1`] = `
Array [
  1,
  2,
  3,
]
`;

exports[`RxJS 5 passes non-observables as is 1`] = `
Object {
  "a": 123,
  "b": "hi!",
}
`;

exports[`RxJS 5 strips dollar sign from Observable property names 1`] = `
Object {
  "a": [Function],
}
`;

exports[`RxJS 6 creates FuncSubject-like action 1`] = `
Array [
  1,
  2,
  3,
]
`;

exports[`RxJS 6 passes non-observables as is 1`] = `
Object {
  "a": 123,
  "b": "hi!",
}
`;

exports[`RxJS 6 strips dollar sign from Observable property names 1`] = `
Object {
  "a": [Function],
}
`;
